File: src/main/scala/kim/dencka/scala/easy/tree/MergeTwoBinaryTrees.scala
package kim.dencka.scala.easy.tree
import scala.util.Random

object MergeTwoBinaryTrees extends App {

  class TreeNode(
    val value: Int,
    val left: TreeNode = null,
    val right: TreeNode = null
  )

  val bt1 =
    new TreeNode(
      1,
      new TreeNode(3, new TreeNode(5)),
      new TreeNode(2)
    )

  val bt2 =
    new TreeNode(
      2,
      new TreeNode(1, right = new TreeNode(4)),
      new TreeNode(3, right = new TreeNode(7))
    )

  val bt3 =
    new TreeNode(
      1,
      right = new TreeNode(
        1,
        right = new TreeNode(
          1,
          right = new TreeNode(
            1,
            right = new TreeNode(
              1,
              right = new TreeNode(
                1,
                new TreeNode(2)
              )
            )
          )
        )
      )
    )

  val bt4 =
    new TreeNode(
      1,
      right = new TreeNode(
        1,
        right = new TreeNode(
          1,
          right = new TreeNode(
            1,
            right = new TreeNode(
              1,
              right = new TreeNode(
                1,
                right = new TreeNode(
                  1,
                  right = new TreeNode(
                    1,
                    right = new TreeNode(
                      1,
                      right = new TreeNode(
                        1,
                        right = new TreeNode(1, new TreeNode(2))
                      )
                    )
                  )
                )
              )
            )
          )
        )
      )
    )

  def mergeTrees(root1: TreeNode, root2: TreeNode): TreeNode =
    if (root1 == null) root2
    else if (root2 == null) root1
    else new TreeNode(
      value = root1.value + root2.value,
      left = mergeTrees(root1.left, root2.left),
      right = mergeTrees(root1.right, root2.right)
    )



  println(mergeTrees(bt3, bt4))

}

File: src/main/scala/kim/dencka/java/easy/MergeTwoBinaryTrees.java
package kim.dencka.java.easy;

import kim.dencka.java.payloads.TreeNode;

class MergeTwoBinaryTrees {
    public static void main(String[] args) {
        TreeNode bt1 = new TreeNode(1);
        TreeNode bt2 = new TreeNode(2);
        bt1.left = new TreeNode(3);
        bt1.left.left = new TreeNode(5);
        bt1.right = new TreeNode(2);
        bt2.left = new TreeNode(1);
        bt2.left.right = new TreeNode(4);
        bt2.right = new TreeNode(3);
        bt2.right.right = new TreeNode(7);

        mergeTrees(bt1, bt2).printAll();

    }

    public static TreeNode mergeTrees(TreeNode root1, TreeNode root2) {
        if (root1 == null) return root2;
        if (root2 == null) return root1;
        root1.val += root2.val;
        root1.left = mergeTrees(root1.left, root2.left);
        root1.right = mergeTrees(root1.right, root2.right);
        return root1;
    }

    // Iterative
//    public static TreeNode mergeTrees(TreeNode t1, TreeNode t2) {
//        if (t1 == null) return t2;
//        Stack<TreeNode[]> stack = new Stack<>();
//        stack.push(new TreeNode[]{t1, t2});
//        TreeNode[] t;
//        while (!stack.isEmpty()) {
//            t = stack.pop();
//            if (t[0] == null || t[1] == null) {
//                continue;
//            }
//            t[0].val += t[1].val;
//            if (t[0].left == null) {
//                t[0].left = t[1].left;
//            } else {
//                stack.push(new TreeNode[]{t[0].left, t[1].left});
//            }
//            if (t[0].right == null) {
//                t[0].right = t[1].right;
//            } else {
//                stack.push(new TreeNode[]{t[0].right, t[1].right});
//            }
//        }
//        return t1;
//    }

}

